src/mlpack/methods/hmm/hmm_impl.hpp

```diff
@@ -406,6 +418,8 @@ void HMM<Distribution>::Generate(const size_t length,
   // distribution of emissions for our starting state.
   dataSequence.col(0) = emission[startState].Random();
 
+  ConvertToLogSpace();
```

Comment:
So, my understanding is that we have to call `ConvertToLogSpace()` before we ever use `logInitial` or `logTransition`, because a user may have modified `initialProxy` or `transitionProxy` by a call to `Initial()` or `Transition()`.  Correct me if I'm wrong on that. :)

If that's correct above, I noticed an additional optimization we can do.  Currently, we compute `log(initialProxy)` and `log(transitionProxy)` every time we call `Predict()`, `Generate()`, or other functions that use the HMM.  But, we only need to do this when we know that the `initialProxy` or `transitionProxy` values could change.  So, what if we add two `bool`s to the class... `recalculateInitial` and `recalculateTransition`.  Then, we can adapt the non-const versions of `Initial()` and `Transition()`:

```
arma::mat& Initial() { recalculateInitial = true; return initialProxy; }
arma::mat& Transition() { recalculateTransition = true; return transitionProxy; }
```

Then we can check in `ConvertToLogSpace()` whether `calculateTransition` or `calculateInitial` are `true`, and do the recomputation only if needed.  This would mean there would only be overhead if the user is ever calling `Initial()` or `Transition()` to modify the matrix.

Reply:
That's a good suggestion but it avoids only some of the over head not all of them. Please note that if the HMM object is non-constant any call to Initial() and Transition() even for reading purpose only, will utilize the non-constant version no matter what.

`auto initial = hmm.Initial()` will call non-constant version of `Initial()` if `hmm` is non-constant. At lease this is the behavior with gcc in my machine (Debian 10)

Reply:
That's true, so you're right that not all overhead will be avoided, but I think that `const auto initial = hmm.Initial()` (or maybe `const auto&`?) will use the const version.  And `const arma::mat& initial = hmm.Initial()` should definitely use the const version.